c10/test/util/complex_test_common.h

```diff
@@ -0,0 +1,464 @@
+  static_assert(std::get<c10::complex<float>>(tup).real() == float(1), "");
+  static_assert(std::get<c10::complex<float>>(tup).imag() == float(2), "");
+}
+
+#if defined(__CUDACC__) || defined(__HIPCC__)
+C10_HOST_DEVICE std::tuple<c10::complex<double>, c10::complex<float>> one_two_thrust() {
+  thrust::complex<float> src(1, 2);
+  c10::complex<double> ret0;
+  c10::complex<float> ret1;
+  ret0 = ret1 = src;
+  return std::make_tuple(ret0, ret1);
+}
+
+TEST(TestAssignment, FromThrust) {
+  auto tup = one_two_thrust();
+  ASSERT_EQ(std::get<c10::complex<double>>(tup).real(), double(1));
+  ASSERT_EQ(std::get<c10::complex<double>>(tup).imag(), double(2));
+  ASSERT_EQ(std::get<c10::complex<float>>(tup).real(), float(1));
+  ASSERT_EQ(std::get<c10::complex<float>>(tup).imag(), float(2));
+}
+#endif
+
+} // namespace assignment
+
+namespace literals {
+
+MAYBE_GLOBAL void test_complex_literals() {
+  using namespace c10::complex_literals;
+  static_assert(std::is_same<decltype(0.5_if), c10::complex<float>>::value, "");
+  static_assert((0.5_if).real() == float(), "");
+  static_assert((0.5_if).imag() == float(0.5), "");
+  static_assert(std::is_same<decltype(0.5_id), c10::complex<double>>::value, "");
+  static_assert((0.5_id).real() == float(), "");
+  static_assert((0.5_id).imag() == float(0.5), "");
+
+  static_assert(std::is_same<decltype(1_if), c10::complex<float>>::value, "");
+  static_assert((1_if).real() == float(), "");
+  static_assert((1_if).imag() == float(1), "");
+  static_assert(std::is_same<decltype(1_id), c10::complex<double>>::value, "");
+  static_assert((1_id).real() == double(), "");
+  static_assert((1_id).imag() == double(1), "");
+}
+
+} // namespace literals
+
+namespace real_imag {
+
+template<typename scalar_t>
+constexpr c10::complex<scalar_t> zero_one() {
+  c10::complex<scalar_t> result;
+  result.imag(scalar_t(1));
+  return result;
+}
+
+template<typename scalar_t>
+constexpr c10::complex<scalar_t> one_zero() {
+  c10::complex<scalar_t> result;
+  result.real(scalar_t(1));
+  return result;
+}
+
+MAYBE_GLOBAL void test_real_imag_modify() {
+  static_assert(zero_one<float>().real() == float(0), "");
+  static_assert(zero_one<float>().imag() == float(1), "");
+  static_assert(zero_one<double>().real() == double(0), "");
+  static_assert(zero_one<double>().imag() == double(1), "");
+
+  static_assert(one_zero<float>().real() == float(1), "");
+  static_assert(one_zero<float>().imag() == float(0), "");
+  static_assert(one_zero<double>().real() == double(1), "");
+  static_assert(one_zero<double>().imag() == double(0), "");
+}
+
+} // namespace real_imag
+
+namespace arithmetic_assign {
+
+template<typename scalar_t>
+constexpr c10::complex<scalar_t> p(scalar_t value) {
+  c10::complex<scalar_t> result(scalar_t(2), scalar_t(2));
+  result += value;
+  return result;
+}
+
+template<typename scalar_t>
+constexpr c10::complex<scalar_t> m(scalar_t value) {
+  c10::complex<scalar_t> result(scalar_t(2), scalar_t(2));
```

Comment:
I think should we just pass in another var `scalar_t res_val` from `test_arithmetic_assign_scalar` instead of defining it in `p` `m` `t` `d` methods

Reply:
I don't understand what do you mean? But I agree this is ugly. I blame this to nvcc because `a += b` should be treated as `constexpr`, but nvcc doesn't.